mifosng-provider/src/main/java/org/mifosplatform/portfolio/loanaccount/domain/Loan.java

```diff
@@ -5615,6 +5640,75 @@ public void addTrancheLoanCharge(Charge charge) {
             trancheCharges.add(new LoanTrancheCharge(charge, this));
         }
     }
+    
+    public List<LoanTransaction> undoLastDisbursal(ScheduleGeneratorDTO scheduleGeneratorDTO, List<Long> existingTransactionIds,
+			List<Long> existingReversedTransactionIds, AppUser currentUser, Map<String, Object> changes, Loan loan) {
+
+
+		validateAccountStatus(LoanEvent.LOAN_DISBURSAL_UNDOLAST);
+
+        List<LoanTransaction> reversedTransactions = new ArrayList<>();
+        validateActivityNotBeforeClientOrGroupTransferDate(LoanEvent.LOAN_DISBURSAL_UNDOLAST, getDisbursementDate());
+        LocalDate actualDisbursementDate = null;
+    	LocalDate lastTransactionDate = getDisbursementDate();
+    	List<LoanTransaction> loanTransactions = retreiveListOfTransactionsExcludeAccruals();
```

Comment:
what exactly you are trying to reverse. are you allowing undo even after some repayment and waivers?